### PR TITLE
Add Wrw tuning

### PR DESCRIPTION
**Pull request by zjing14**
- Rework Wrw cpp
- Add Wrw tuning
- Add Wrw tuning test

### CODE REVIEW

File: mlir/include/mlir/Target/MIOpenCPP.h [hunk 1]
```diff
@@ -204,6 +182,41 @@ class PopulateParamsBase {
       } else {
         input2GemmKVectorizable = false;
       }
+    } else if (opType == mlir::miopen::ConvOpType::Conv2DBwdWeightOpType) {
+      // For input tensor
+      // currently, fix that GemmK (NHiWi) is always vectorizable
```
Comment (whchung) [inline mlir/include/mlir/Target/MIOpenCPP.h:187]: @zjing14  want to check if this comment is correct. what's GemmK and GemmN on backward weight?

Comment (zjing14) [reply mlir/include/mlir/Target/MIOpenCPP.h:187]: GemmM = K
GemmK = NHW
GemmN = CYX

**Review by whchung — APPROVED** [review]

File: mlir/include/mlir/Target/MIOpenCPP.h [hunk 2]
```diff
@@ -269,6 +297,8 @@ class PopulateParamsBase {
       obtainOutputVecLen(dimIndexVal, vecLen);
     } else if (opType == mlir::miopen::ConvOpType::Conv2DBwdDataOpType) {
       obtainInputVecLen(dimIndexVal, vecLen);
+    } else if (opType == mlir::miopen::ConvOpType::Conv2DBwdWeightOpType) {
+      obtainInputVecLen(dimIndexVal, vecLen);
```
Comment (jerryyin) [inline mlir/include/mlir/Target/MIOpenCPP.h:301]: In bww, the gemmC should be Filter

Comment (jerryyin) [inline mlir/include/mlir/Target/MIOpenCPP.h:301]: This needs to be `obtainFilterVecLen`

File: mlir/test/Dialect/MIOpen/translate_cflags_wrw.mlir [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+// CHECK: -DCK_PARAM_TUNABLE_GEMM_M_LEVEL1_CLUSTER=4
+// CHECK: -DCK_PARAM_TUNABLE_GEMM_M_PER_BLOCK=128
+// CHECK: -DCK_PARAM_TUNABLE_GEMM_M_PER_THREAD=4
+// CHECK: -DCK_PARAM_TUNABLE_GEMM_N_LEVEL0_CLUSTER=4
+// CHECK: -DCK_PARAM_TUNABLE_GEMM_N_LEVEL1_CLUSTER=4
+// CHECK: -DCK_PARAM_TUNABLE_GEMM_N_PER_BLOCK=128
+// CHECK: -DCK_PARAM_TUNABLE_GEMM_N_PER_THREAD=4
+// CHECK: -DCK_THREADWISE_GEMM_USE_AMD_INLINE_ASM=1
+// CHECK: -std=c++14
+// CHECK: -D__HIP_PLATFORM_HCC__=1
```
Comment (jerryyin) [inline mlir/test/Dialect/MIOpen/translate_cflags_wrw.mlir:99]: I don't see the `-DCK_PARAM_PROBLEM_CONV_DIRECTION_BACKWARD_WEIGHT=1` set, do we need it in wrw?

Comment (zjing14) [reply mlir/test/Dialect/MIOpen/translate_cflags_wrw.mlir:99]: We have a dedicated wrw kernel, so DCK_PARAM_PROBLEM_CONV_DIRECTION_BACKWARD_WEIGHT is not necessary. DCK_PARAM_PROBLEM_CONV_DIRECTION_BACKWARD_WEIGHT is used for reusing fwd kernel for wrw.

Comment (jerryyin) [reply mlir/test/Dialect/MIOpen/translate_cflags_wrw.mlir:99]: Sounds good thanks for clarification.

**Comment by jerryyin** [comment]
Fixed in #24

**Pull request merged**